File: internal/service/realtime.go
package service

import "time"

type RealtimePublisher interface {
	PublishLocation(event LocationEvent)
	PublishDeviceStatus(event DeviceStatusEvent)
	PublishAlarm(event AlarmEvent)
}

type LocationEvent struct {
	DeviceSN     string    `json:"device_sn"`
	TopicPrefix  string    `json:"topic_prefix"`
	Latitude     float64   `json:"lat"`
	Longitude    float64   `json:"lng"`
	Time         time.Time `json:"time"`
	StillSeconds int       `json:"still_seconds"`
	GPSState     string    `json:"gps_state"`
	Status       int       `json:"status"`
}

type DeviceStatusEvent struct {
	DeviceSN    string     `json:"device_sn"`
	TopicPrefix string     `json:"topic_prefix"`
	Status      int        `json:"status"`
	GPSState    string     `json:"gps_state"`
	Battery     int        `json:"battery"`
	IMEI        string     `json:"imei,omitempty"`
	ICCID       string     `json:"iccid,omitempty"`
	LastOnline  *time.Time `json:"last_online,omitempty"`
	LastFixAt   *time.Time `json:"last_fix_at,omitempty"`
}

type AlarmEvent struct {
	DeviceSN  string    `json:"device_sn"`
	Type      string    `json:"type"`
	Content   string    `json:"content"`
	CreatedAt time.Time `json:"created_at"`
}

File: web/src/types/realtime.ts
export interface LocationEvent {
  device_sn: string;
  topic_prefix: string;
  lat: number;
  lng: number;
  time: string;
  still_seconds: number;
  gps_state: string;
  status: number;
}

export interface DeviceStatusEvent {
  device_sn: string;
  topic_prefix: string;
  status: number;
  gps_state: string;
  battery: number;
  imei?: string;
  iccid?: string;
  status_payload?: Record<string, unknown>;
  config_payload?: Record<string, unknown>;
  status_updated_at?: string;
  config_updated_at?: string;
  last_online?: string;
  last_fix_at?: string;
}

export interface AlarmEvent {
  device_sn: string;
  type: string;
  content: string;
  created_at: string;
}

export type RealtimeEnvelope =
  | { type: "location"; data: LocationEvent }
  | { type: "device_status"; data: DeviceStatusEvent }
  | { type: "alarm"; data: AlarmEvent };
